streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java

```diff
@@ -730,16 +827,150 @@ private static void buildAssignedActiveTaskAndPartitionsList(final List<TaskId>
+                if (assignments.get(consumer).size() >= maxTasksPerClient) {
+                    return Collections.emptyMap();
+                }
+
+                assignments.get(consumer).add(task);
+
+                // If we have now reached capacity, remove it from set of consumers who still need more tasks
+                if (assignments.get(consumer).size() == maxTasksPerClient) {
+                    unfilledConsumers.remove(consumer);
+                }
+            }
+        }
+
+        // Interleave any remaining tasks among the consumers with remaining capacity
+        Collections.sort(newTasks);
+        while (!newTasks.isEmpty()) {
+            if (unfilledConsumers.isEmpty()) {
+                throw new IllegalStateException("Some tasks could not be distributed");
+            }
+
+            final Iterator<String> consumerIt = unfilledConsumers.iterator();
+            // Loop through the unfilled consumers and distribute tasks
+            while (consumerIt.hasNext()) {
+                final String consumer = consumerIt.next();
+                final List<TaskId> consumerAssignment = assignments.get(consumer);
+                final TaskId task = newTasks.poll();
+                if (task == null) {
+                    break;
+                }
+
+                consumerAssignment.add(task);
+                if (consumerAssignment.size() == maxTasksPerClient) {
+                    consumerIt.remove();
+                }
+            }
+        }
+
+        return assignments;
+    }
+
+    /**
+     * Get the previous consumer for the partitions of a task
+     *
+     * @param taskPartitions the TopicPartitions for a single given task
+     * @param clientOwnedPartitions the partitions owned by all consumers in a client
+     * @param allOwnedPartitions all partitions claimed as owned by any consumer in any client
+     * @return set of consumer(s) that previously owned the partitions in this task
+     *         empty set signals that it is a new task, or its previous owner is no longer in the group
+     */
+    Set<String> previousConsumersOfTaskPartitions(final Set<TopicPartition> taskPartitions,
+                                                 final Map<TopicPartition, String> clientOwnedPartitions,
+                                                 final Set<TopicPartition> allOwnedPartitions) {
+        // this "consumer" indicates a partition was owned by someone from another client -- we don't really care who
+        final String foreignConsumer = "";
+        final Set<String> previousConsumers = new HashSet<>();
+
+        for (final TopicPartition tp : taskPartitions) {
+            final String currentPartitionConsumer = clientOwnedPartitions.get(tp);
+            if (currentPartitionConsumer != null) {
+                previousConsumers.add(currentPartitionConsumer);
+            } else if (allOwnedPartitions.contains(tp)) {
```

Comment:
If this partition is owned by a consumer in this client, then `currentPartitionConsumer != null` and `currentPartitionConsumer` gets added to the `previousConsumers` set. We don't even check `allOwnedPartitions` in that case since it's in an "else if"